### ISSUE TITLE
Unable to produce a successfull Magisk install on AVD Manager created emulators 

### ISSUE THREAD

**Issue by SuspensionPoint** [shakalaca/MagiskOnEmulator#15]
I'm running Windows 10.

I've tried using the Pixel 3a XL emulator with the following API/Targets:

- API: 30 Target: Android 11.0 (Google APIs) x86
- API 29 Target: Android 10 (Google APIs) x86

Following the steps in the read me to install the Canary build results in the following:

A fully working emulator with Magisk Manager installed, but Magisk not installed.
![wack](https://user-images.githubusercontent.com/62025450/98019745-c114df00-1dc7-11eb-9a49-6cc488ba093a.JPG)

This isn't Canary specific from what I can tell. I haven't been able to succeed when not using the canary flag.

My expectation of this script is to run `./path.bat canary` on a the fresh ramdisk.img and end up with a fully rooted android emulator w/ Magisk installed and Magisk Manager.

Is this the goal of the script? Is that what the expected behavior should be, or is it possible that I am missing a step?

Also, thank you guys for working on this project! This process has always been an absolute pain in the ass. I appreciate it!

EDIT:
I just ran through the ReadMe again on a Pixel 3a XL Android 11.0 API 30 x86 emulator and ended up with the same result. I've went through the process probably 15 times now and haven't ended up with a successful install.

Best wishes!

**Comment by fusionneur** [comment]
Do you still have the log ?

**Comment by fusionneur** [comment]
I tested and was able to root on Pixel 3a XL API 29 x86 (Google API)

[img]https://i.imgur.com/6RjJIsK.png[/img]

**Comment by fusionneur** [comment]
As I was commenting on the pull request to modify the readme file: "Alternatively, you can execute `patch.sh canary` or `patch.bat canary` to install latest canary Magisk on the ramdisk.img. **This requires AVD internet connectivity towards github.**"

Do you have internet connectivity inside AVD?

**Comment by SuspensionPoint** [comment]
> 
> 
> Do you still have the log ?

Which log? Are you asking for the console output of the .patch.bat? If so, here:

> PS C:\projects\MagiskOnEmulator> .\patch.bat canary
ramdisk.img: 1 file pushed. 16.1 MB/s (1475652 bytes in 0.087s)
magisk_emu.zip: 1 file pushed. 50.9 MB/s (6196156 bytes in 0.116s)
update-binary: 1 file pushed. 47.4 MB/s (3527756 bytes in 0.071s)
process.sh: 1 file pushed. 1.9 MB/s (5782 bytes in 0.003s)
[*] Extracting busybox ..
[*] Checking Android version
[*] Fetching canary version of Magisk ..
Connecting to raw.githubusercontent.com (151.101.0.133:443)
wget: note: TLS certificate validation not implemented
saving to 'magisk.zip'
magisk.zip             0% |                                | 12339  0:16:10 ETA
magisk.zip            39% |************                    | 4615k  0:00:03 ETA
magisk.zip            85% |***************************     | 9991k  0:00:00 ETA
wget: got bad TLS record (len:2019) while expecting encrypted data
 03 03 40 18 c0 0e f5 db bd 12 e3 0d 90 86 0b e9 f4 b8 ee 92 77 cb 40 b3 b4 b3 f8 ef 7e 82 d4 c7 fd 51 38 70 2a 5c 7d cc 96 cb 2f 0c 25 bd 7d 72 50 81 18 fb ac 46 c4 c8 45 0d 61 34 29 86 57 5e f5 52 48 89 5d 7e 18 5e c9 19 08 88 82 dd 56 28 5d 10 9e 9d ff 9b f7 57 76 ac 75 87 cc 9a c6 66 4e 10 d8
wget: connection closed prematurely
[*] Failed to get full magisk.zip, retry .. ( 2683 )
Connecting to raw.githubusercontent.com (151.101.0.133:443)
wget: note: TLS certificate validation not implemented
saving to 'magisk.zip'
wget: got bad TLS record (len:2019) while expecting encrypted data
 03 03 40 18 d3 84 48 00 26 75 f2 c2 4c 8e 97 cc d4 d3 34 20 85 34 e4 67 a7 12 07 9b 10 f5 0f 96 6d 3c 90 90 e0 d3 64 0e 88 47 94 c8 9c fb b0 0b d5 49 65 27 24 73 9a c3 09 47 75 63 93 20 ea 43 98 30 1c 8c b6 b2 81 23 bd e6 4f 2b e6 7b 67 4e 2d 39 b2 f6 9b 67 04 5b 77 98 6f e9 7d d6 fc cd 44 80 b1
wget: connection closed prematurely
[*] Failed to get full magisk.zip, retry .. ( 2901 )
Connecting to raw.githubusercontent.com (151.101.0.133:443)
wget: note: TLS certificate validation not implemented
saving to 'magisk.zip'
magisk.zip           100% |********************************| 11.3M  0:00:00 ETA
'magisk.zip' saved
wget: short read, have only 30
[-] API level greater then 30
[*] Check if we need to repack ramdisk before patching ..
[-] Multiple cpio archives detected
[*] Unpacking ramdisk ..
[-] Dumping from 0 to 2184316 ..
[-] Dumping from 2184448 to 4313260 ..
[*] Repacking ramdisk ..
[*] Unzipping Magisk ..
[*] Checking ramdisk status ..
[-] Stock boot image detected
[*] Patching ramdisk ..
Loading cpio: [/data/local/tmp/ramdisk.img]
Create directory [.backup] (0000)
Move [init] -> [.backup/init]
Dump cpio: [/data/local/tmp/ramdisk.img]
Loading cpio: [/data/local/tmp/ramdisk.img]
Create directory [apex] (0755)
Dump cpio: [/data/local/tmp/ramdisk.img]
Loading cpio: [/data/local/tmp/ramdisk.img]
Add entry [init] (0750)
Add entry [.backup/.magisk] (0000)
Dump cpio: [/data/local/tmp/ramdisk.img]
Loading cpio: [/data/local/tmp/ramdisk.img]
Patch with flag KEEPVERITY=[false] KEEPFORCEENCRYPT=[true]
Found fstab file [first_stage_ramdisk/fstab.ranchu]
Remove pattern [,avb=vbmeta]
Found fstab file [fstab.ranchu]
Remove pattern [,avb=vbmeta]
Dump cpio: [/data/local/tmp/ramdisk.img]
[*] Done patching, compressing ramdisk ..
[*] Installing MagiskManager ..
[*] Installing su binaries ..
[*] Clean up
/data/local/tmp/ramdisk.img: 1 file pulled. 47.7 MB/s (1810349 bytes in 0.036s)
PS C:\projects\MagiskOnEmulator> adb emu kill

> 
> 
> As I was commenting on the pull request to modify the readme file: "Alternatively, you can execute `patch.sh canary` or `patch.bat canary` to install latest canary Magisk on the ramdisk.img. **This requires AVD internet connectivity towards github.**"
> 
> Do you have internet connectivity inside AVD?

I do believe so. I've been connecting perfectly fine through the emulator itself.

Is there a chance that there is a difference in the .bat vs .sh scripts?

**Comment by SuspensionPoint** [comment]
Just in case, here are the details of my Pixel 3a XL API 29 x86 (Google API) instance:

> Name: root
> 
> CPU/ABI: Google APIs Intel Atom (x86)
> 
> Path: C:\Users\\{name}\\.android\avd\root.avd
> 
> Target: google_apis [Google APIs] (API level 29)
> 
> Skin: pixel_3_xl
> 
> SD Card: 512M
> 
> fastboot.chosenSnapshotFile: 
> 
> runtime.network.speed: full
> 
> hw.accelerometer: yes
> 
> hw.device.name: pixel_3_xl
> 
> hw.lcd.width: 1440
> 
> hw.initialOrientation: Portrait
> 
> image.androidVersion.api: 29
> 
> tag.id: google_apis
> 
> hw.mainKeys: no
> 
> hw.camera.front: emulated
> 
> avd.ini.displayname: root
> 
> hw.gpu.mode: auto
> 
> hw.ramSize: 1536
> 
> PlayStore.enabled: false
> 
> fastboot.forceColdBoot: no
> 
> hw.cpu.ncore: 2
> 
> hw.keyboard: yes
> 
> hw.sensors.proximity: yes
> 
> hw.dPad: no
> 
> hw.lcd.height: 2960
> 
> vm.heapSize: 256
> 
> skin.dynamic: yes
> 
> hw.device.manufacturer: Google
> 
> hw.gps: yes
> 
> hw.audioInput: yes
> 
> image.sysdir.1: system-images\android-29\google_apis\x86\
> 
> showDeviceFrame: yes
> 
> hw.camera.back: virtualscene
> 
> AvdId: root
> 
> hw.lcd.density: 560
> 
> hw.arc: false
> 
> hw.device.hash2: MD5:9bbaa408d4a8691920c4057ef7844f35
> 
> fastboot.forceChosenSnapshotBoot: no
> 
> fastboot.forceFastBoot: yes
> 
> hw.trackBall: no
> 
> hw.battery: yes
> 
> hw.sdCard: yes
> 
> tag.display: Google APIs
> 
> runtime.network.latency: none
> 
> disk.dataPartition.size: 800M
> 
> hw.sensors.orientation: yes
> 
> avd.ini.encoding: UTF-8
> 
> hw.gpu.enabled: yes

**Comment by fusionneur** [comment]
"Is there a chance that there is a difference in the .bat vs .sh scripts?" 

I tested with the .bat via powershell

**Comment by fusionneur** [comment]
Does it always fail ? Did you retry?

**Comment by fusionneur** [comment]
I think it is related with the issue I also reported a while ago #13 and the previous one #8 

For some reason android and wget is buggy..

**Comment by SuspensionPoint** [comment]
> 
> 
> I think it is related with the issue I also reported a while ago #13 and the previous one #8 (which was fixed with retry warkaround)
> 
> For some reason android and wget is buggy..

Interesting.

I've retried the whole process 10-15 times now. Do you think I should just try re-running the .bat script on the ramdisk.img immediately after the wget failing? Or do what I've been doing, which is replace the ramdisk.img with a clean one.

I've never been able to get a successful run of the script without the wget/TLS complaint.

**Comment by fusionneur** [comment]
"I've retried the whole process 10-15 times now."
Oh 10-15 is quite a lot 👍 Then i believe you encounter a different issue. For me the bug #13 is quite rare

"Do you think I should just try re-running the .bat script on the ramdisk.img immediately after the wget failing? Or do what I've been doing, which is replace the ramdisk.img with a clean one."
No, the bat should be run only once. For fresh patching, yes, you need to overwrite the ramdisk with the original one.

"I've never been able to get a successful run of the script without the wget/TLS complaint."
Tls complaints are normal and were fixed in #8 by retrying and reconstructing the zip.

**Comment by SuspensionPoint** [comment]
I'm trying the process one more time with a fresh setup.

This is my output:

> 
> PS C:\projects\MagiskOnEmulator> .\patch.bat canary
> adb server version (41) doesn't match this client (39); killing...
> * daemon started successfully *
> ramdisk.img: 1 file pushed. 16.6 MB/s (899609 bytes in 0.052s)
> magisk_emu.zip: 1 file pushed. 68.5 MB/s (6196156 bytes in 0.086s)
> update-binary: 1 file pushed. 18.4 MB/s (3527756 bytes in 0.183s)
> process.sh: 1 file pushed. 2.2 MB/s (5782 bytes in 0.002s)
> [*] Extracting busybox ..
> [*] Checking Android version
> [*] Fetching canary version of Magisk ..
> Connecting to raw.githubusercontent.com (151.101.0.133:443)
> wget: note: TLS certificate validation not implemented
> saving to 'magisk.zip'
> magisk.zip            12% |****                            | 1459k  0:00:06 ETA
> magisk.zip            29% |*********                       | 3459k  0:00:04 ETA
> magisk.zip            48% |***************                 | 5699k  0:00:03 ETA
> magisk.zip            68% |*********************           | 7955k  0:00:01 ETA
> magisk.zip            86% |***************************     |  9.8M  0:00:00 ETA
> magisk.zip           100% |********************************| 11.3M  0:00:00 ETA
> 'magisk.zip' saved
> [*] Unzipping Magisk ..
> [*] Checking ramdisk status ..
> [-] Stock boot image detected
> [*] Patching ramdisk ..
> 
> Loading cpio: [/data/local/tmp/ramdisk.img]
> Create directory [.backup] (0000)
> Move [init] -> [.backup/init]
> Dump cpio: [/data/local/tmp/ramdisk.img]
> Loading cpio: [/data/local/tmp/ramdisk.img]
> Add entry [init] (0750)
> Add entry [.backup/.magisk] (0000)
> Dump cpio: [/data/local/tmp/ramdisk.img]
> Loading cpio: [/data/local/tmp/ramdisk.img]
> Patch with flag KEEPVERITY=[false] KEEPFORCEENCRYPT=[true]
> Found fstab file [fstab.ranchu]
> Remove pattern [,avb=vbmeta]
> Dump cpio: [/data/local/tmp/ramdisk.img]
> [*] Done patching, compressing ramdisk ..
> [*] Installing MagiskManager ..
> [*] Installing su binaries ..
> [*] Clean up
> /data/local/tmp/ramdisk.img: 1 file pulled. 68.2 MB/s (1241574 bytes in 0.017s)

One of the last steps mentions copying over the patched ramdisk.img "back to the AVD directory".

Is it supposed to be copy over to the directory such as:

> C:\Users\\{name}\\AppData\Local\Android\sdk\system-images\android-29\google_apis\x86


Or the directory of the actual AVD such as:

> C:\Users\\{name}\\.android\avd\root.avd

Perhaps its an issue with my ADB Version?

>  adb version
> Android Debug Bridge version 1.0.39
> Revision 3db08f2c6889-android
> Installed as C:\Program Files (x86)\Essential\ADB\adb.exe

**Comment by fusionneur** [comment]
I believe you replaced the ramdisk.img from: C:\Users\{user}\AppData\Local\Android\Sdk\system-images\android-29\google_apis\x86 with the patched ramdisk.img from the , right? (Asking just to be sure)

**Comment by SuspensionPoint** [comment]
> 
> 
> I believe you replaced the ramdisk.img from: C:\Users{user}\AppData\Local\Android\Sdk\system-images\android-29\google_apis\x86 with the patched ramdisk.img from the , right? (Asking just to be sure)

Yeah. That's the way I have been doing it. 

I believe my next step is to update ADB.

**Comment by fusionneur** [comment]
I see you already posted the steps you followed.

"Is it supposed to be copy over to the directory such as:

C:\Users\{name}\AppData\Local\Android\sdk\system-images\android-29\google_apis\x86"

Yes you need to overwrite the original ramdisk, so that it cold boots with the patched one.

**Comment by fusionneur** [comment]
I don't think it is related with ADB version..

**Comment by SuspensionPoint** [comment]
I'm certainly stumped if it's not some odd version issue then.

It definitely appears that I'm following the steps correctly at this point, but I just don't get the same result.

**Comment by fusionneur** [comment]
Does it work for you without 'canary' ? Also does it work for you on Android 9 for example?

**Comment by SuspensionPoint** [comment]
I'll try a run run through with Android 9 without canary. I'll update w/ the results.

Excuse my ignorance, but is it possible that I'm missing some necessary SDK tools installed? Such as Google Play Services?

**Comment by fusionneur** [comment]
I took a second look on your log and indeed it says something related to adb version 
"adb server version (41) doesn't match this client (39); killing..."

You can use adb which comes with Android Studio. You already have it. Path would be something like this:
C:\Users\Fusion\AppData\Local\Android\Sdk\platform-tools

You can add the path to win env variables to be able to use adb from anywhere.


Related to the other question,
"Excuse my ignorance, but is it possible that I'm missing some necessary SDK tools installed? Such as Google Play Services?"
Patching ramdisk works for any android image (with or without google api or play store)

**Comment by SuspensionPoint** [comment]
I had some issues with my Environment Variables and adb version. I got all of that fixed up and updated. I tried the process on Android 9 w/ no canary. Still didn't have Magisk installed.

I then retried the process on Pixel 3a XL API 29 x86 (Google API). 

> 
> PS C:\projects\MagiskOnEmulator> .\patch.bat canary                                                                     ramdisk.img: 1 file pushed, 0 skipped. 215.2 MB/s (899609 bytes in 0.004s)
> magisk_emu.zip: 1 file pushed, 0 skipped. 883.8 MB/s (6196156 bytes in 0.007s)
> update-binary: 1 file pushed, 0 skipped. 977.5 MB/s (3527756 bytes in 0.003s)
> process.sh: 1 file pushed, 0 skipped. 14.3 MB/s (5782 bytes in 0.000s)
> [*] Extracting busybox ..
> [*] Checking Android version
> [*] Fetching canary version of Magisk ..
> Connecting to raw.githubusercontent.com (151.101.0.133:443)
> wget: note: TLS certificate validation not implemented
> saving to 'magisk.zip'
> magisk.zip            14% |****                            | 1638k  0:00:06 ETA
> wget: got bad TLS record (len:2019) while expecting encrypted data
>  03 03 40 18 7e 50 14 8f 02 9d f6 17 43 93 26 67 14 70 0f 71 80 34 00 0c 22 0c dd 38 09 36 71 5e 05 16 5f 69 da d1 a7 d2 17 37 e5 96 f6 27 07 42 2c 2d 2c 25 5f 23 88 ce f1 c5 00 18 68 f5 d5 df 68 f5 49 ed cc f7 3d 41 de 37 70 0b 5d 2a 98 aa 91 1e 96 71 89 e1 62 73 d5 b1 61 36 61 e2 f8 fd 04 58 d8
> wget: connection closed prematurely
> [*] Failed to get full magisk.zip, retry .. ( 2886 )
> Connecting to raw.githubusercontent.com (151.101.0.133:443)
> wget: note: TLS certificate validation not implemented
> saving to 'magisk.zip'
> magisk.zip           100% |********************************| 11.3M  0:00:00 ETA
> 'magisk.zip' saved
> [*] Unzipping Magisk ..
> [*] Checking ramdisk status ..
> [-] Stock boot image detected
> [*] Patching ramdisk ..
> 
> Loading cpio: [/data/local/tmp/ramdisk.img]
> Create directory [.backup] (0000)
> Move [init] -> [.backup/init]
> Dump cpio: [/data/local/tmp/ramdisk.img]
> Loading cpio: [/data/local/tmp/ramdisk.img]
> Add entry [init] (0750)
> Add entry [.backup/.magisk] (0000)
> Dump cpio: [/data/local/tmp/ramdisk.img]
> Loading cpio: [/data/local/tmp/ramdisk.img]
> Patch with flag KEEPVERITY=[false] KEEPFORCEENCRYPT=[true]
> Found fstab file [fstab.ranchu]
> Remove pattern [,avb=vbmeta]
> Dump cpio: [/data/local/tmp/ramdisk.img]
> [*] Done patching, compressing ramdisk ..
> [*] Installing MagiskManager ..
> [*] Installing su binaries ..
> [*] Clean up
> /data/local/tmp/ramdisk.img: 1 file pulled, 0 skipped. 78.1 MB/s (1241574 bytes in 0.015s)

![aasdasda](https://user-images.githubusercontent.com/62025450/98034820-35a64880-1ddd-11eb-9e6d-86a8c99a6645.JPG)


Still nothing. No matter what I've done or changed, the Magisk still wont install.

**Comment by SuspensionPoint** [comment]
Here's what I'm doing:

1. Download fresh sdk/system image
2. Copy ramdisk image over to MagiskOnEmulator path.
3. Create emulator.
4. Start Emulator.
5. Run ./patch.bat canary
6. copy ramdisk image back to the original system image path.
7. Turn off emulator.
8. Cold boot emulator.

Magisk manager installed, but magisk install listed as N/A

**Comment by fusionneur** [comment]
What size is your patched ramdisk.img in KB ? Mine is 1213 KB.

**Comment by SuspensionPoint** [comment]
Thank you for all the help!

I side installed Android Studio Canary and used that explicit revision 11 image of Android 10 x86 and it worked!

**Comment by fusionneur** [comment]
Interesting ... by curiosity what version of Android Studio were you using?

**Comment by iotY** [comment]
@fusionneur , `patch.sh canary` works fine but resulting `/data/adb/magisk` is empty, I have to copy manually {x86}magiskboot-init-init64, busybox_x86-selinux and latest util_functions.sh to get modules flash & work. Is this natural AVD heritage behavior?

**Comment by fusionneur** [comment]
Normally you shouldn't perform any other steps other than the ones documented to have magisk working.
It's better to open a separate issue and provide as much info as you can to further look into the issue.

**Comment by iotY** [comment]
Indeed , I probably closed (turn off) emulator prematurely : didn't get this stuff was adb-ed dynamically by patch.sh
Sorry.

**Comment by kryokil** [comment]
patched fine with latest magisk canary and x86_64 on api30 however having problem installing modules. Anybody has similar issue?

**Comment by fusionneur** [comment]
Please open a separate issue with the details and logs

**Comment by kryokil** [comment]
I have just actually done it :). finally. Magisk and EdXposed installed .
